Implement the new Bridge system for the ME Bridge

SirEndii opened the pull request:
A more descriptive description will come and a review is not yet recommend

SirEndii committed:
Bump AP and Neo

SirEndii committed:
Merge branch 'dev/1.21.1-rs2' into dev/1.21.1-mebridge

SirEndii committed:
Port the 0.8 bridge system to 1.21.1 for the me bridge

SirEndii committed:
Fixed external item storage functions for vanilla and most modded containers

SirEndii committed:
Implement full chemical support and improved some legacy features

SirEndii committed:
Add missing null checks

SirEndii committed:
checkstyle

SirEndii committed:
Fixed estimate of used external item storage

SirEndii committed:
Merge branch 'dev/1.21.1-rs2' into dev/1.21.1-mebridge

# Conflicts:
#	src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/MEBridgePeripheral.java
#	src/main/java/de/srendi/advancedperipherals/common/blocks/blockentities/MEBridgeEntity.java

SirEndii committed:
Merge branch 'dev/1.21.1' into dev/1.21.1-mebridge

# Conflicts:
#	src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/MEBridgePeripheral.java

SirEndii committed:
Fixed properties of the crafting job

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/appliedenergistics/AECraftJob.java:
```diff
@@ -0,0 +1,309 @@
+    }
+
+    private CraftingJobStatus getJobStatus() {
+        if (jobStatus == null || jobStatus.get() == null && cachedStatus != null) {
+            return cachedStatus;
+        }
+        cachedStatus = jobStatus.get();
+        return jobStatus.get();
```
```suggestion
        if (cachedStatus == null && jobStatus != null) {
            cachedStatus = jobStatus.get();
        }
        return cachedStatus;
```
I think you mean this

SirEndii replied:
That makes more sense

zyxkad replied:
Actually I'm not sure if `jobStatus.get() == null` is on purpose, like if the cache should be used only after the job is finished or the status can be cached all the time?

SirEndii replied:
I've added a comment which hopefully explains it a bit better

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/appliedenergistics/AECraftJob.java:
```diff
@@ -0,0 +1,309 @@
+        }
+        for (ICraftingCPU cpu : service.getCpus()) {
+            if (cpu instanceof CraftingCPUCluster cpuCluster) {
+                if (cpuCluster.craftingLogic.getLastLink() != null && cpuCluster.craftingLogic.getLastLink().getCraftingID().equals(jobLink.getCraftingID())) {
```
Can we define a temporary variable here instead of copy paste same expr twice 😅

SirEndii replied:
Uh does that even work since the expression can change and the second expression is called later in the supplier?
Or do I overlook something here

zyxkad replied:
I mean
```suggestion
                final ICraftingLink lastLink = cpuCluster.craftingLogic.getLastLink();
                if (lastLink != null && lastLink.getCraftingID().equals(jobLink.getCraftingID())) {
```

SirEndii committed:
Merge branch 'dev/1.21.1' into dev/1.21.1-mebridge

# Conflicts:
#	src/main/java/de/srendi/advancedperipherals/common/addons/APAddons.java
#	src/main/java/de/srendi/advancedperipherals/common/addons/appliedenergistics/AppEngApi.java

SirEndii committed:
Checkstyle fix and added ae2things to APAddon.java

SirEndii committed:
Change the logic of `getJobStatus`

SirEndii committed:
Add `isCrafting` to rs bridge objects

SirEndii committed:
Return an empty stack if there is no stored AEKey while trying to export

SirEndii committed:
Fix chemical used function
Remove the property `powered` of the ME Drive since it would be always true due that it would only return false when the ME System is offline and at this point, the bridge would just return a debug message that the system is offline with the default value. Also, there wouldn't be a RS equivalent

SirEndii committed:
Adapt RS pattern properties to the AE2 ones and vise versa

SirEndii committed:
Checkstyle

SirEndii committed:
map -> properties

SirEndii committed:
Fixed `getPatterns`

SirEndii committed:
Fixed wrong chemical AE2 stack properties

SirEndii committed:
Add missing LuaFunction annotations and rename every `list` function to `get`

SirEndii committed:
AppEngAPI -> AEApi to adapt the class name to the one from RS

SirEndii commented:
@zyxkad Feature wise, well tested in my RS and AE2 world, everything should be pretty the same to RS in terms of properties of the objects and function syntax
The only thing that I am not too happy about is the layout off the AEApi file, but that's nothing I want to bother with currently (and don't have the time to). That's possibly a thing I want to change with 0.8? But tbh, not a priority even with the urgh code quality for me now
I don't see anything feature wise or maybe some potential performance issues that would need a change, but you have better eyes than me and I would love to release the beta of 1.21.1 this or next week

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/appliedenergistics/AEApi.java:
```diff
@@ -0,0 +1,935 @@
+    }
+
+    public static <T extends AEKey> Map<String, Object> parseAeStack(Pair<Long, T> stack, @Nullable ICraftingService service) {
+        if (stack == null || stack.getRight() == null)
```
is stack becomes `null` expected? either way I'd recommend returns null instead of an empty map.

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/appliedenergistics/AEApi.java:
```diff
@@ -0,0 +1,935 @@
+                if (stack.getItem() instanceof ChemicalStorageCell) {
+                    BasicCellInventory cellInventory = BasicCellHandler.INSTANCE.getCellInventory(stack, null);
+
+                    used = cellInventory.getUsedBytes();
```
```suggestion
                    used += cellInventory.getUsedBytes();
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/appliedenergistics/AEApi.java:
```diff
@@ -0,0 +1,935 @@
+                for (int i = 0; i < itemHandler.getSlots(); i++) {
+                    ItemStack stack = itemHandler.getStackInSlot(i);
+
+                    // Use 64 as a rough estimate if no item is in that slot
+                    if (stack.isEmpty()) total += 64;
+
+                    // If the slot is not empty, use the max stack size of that stack
+                    total += stack.getMaxStackSize();
```
```suggestion
                    total += stack.isEmpty() ? itemHandler.getSlotLimit(i) : stack.getMaxStackSize();
```

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/appliedenergistics/MEChemicalHandler.java:
```diff
@@ -0,0 +1,58 @@
+        if(resource.isEmpty())
+            return resource;
+
+        ChemicalStack inserted = resource.copy();
+        long amountInserted = storageMonitor.insert(MekanismKey.of(resource), resource.getAmount(), action == Action.SIMULATE ? Actionable.SIMULATE : Actionable.MODULATE, actionSource);
+        inserted.setAmount(resource.getAmount() - amountInserted);
+        return inserted;
```
```suggestion
        ChemicalStack remain = resource.copy();
        long amountInserted = storageMonitor.insert(MekanismKey.of(resource), resource.getAmount(), action == Action.SIMULATE ? Actionable.SIMULATE : Actionable.MODULATE, actionSource);
        remain.setAmount(resource.getAmount() - amountInserted);
        return remain;
```

SirEndii committed:
Adapt the right default values to the ME Bridge functions

* Return the slot limit of the item handler instead of 64 when the stack is empty

* Some small refactoring

SirEndii committed:
Checkstyle

SirEndii committed:
Merge branch 'dev/1.21.1' into dev/1.21.1-mebridge

SirEndii committed:
reformat

zyxkad commented on src/main/java/de/srendi/advancedperipherals/common/addons/computercraft/peripheral/MEBridgePeripheral.java:
```diff
@@ -130,429 +155,796 @@ protected MethodResult importToME(@NotNull IArguments arguments, @Nullable IItem
+
+        for (AECraftJob job : bridge.getJobs()) {
+            if (job.getId() == id) {
+                foundJob = job;
```
```suggestion
                foundJob = job;
                break;
```

zyxkad reviewed (approved):
otherwise lgtm

SirEndii committed:
Add `break` to improve job search efficiency
